### DISCUSSION TITLE
DPLR: 2 different Wannier centroids ; bond types and fix dplr

### DISCUSSION THREAD

**Discussion by axgmz** [deepmodeling/deepmd-kit#2526] [Q&A] [answered]
Hello!

**Context and models trained:**
I am trying to use DPLR for a slightly more complex system than the example bulk H2O by adding an anion. I have a pretty good dataset with coords, energies, forces and wannier centroids on O and my anion.
A short range DeePMD NNP with this training set is doing good. A DeepDipole model is also good at predicting the Wannier centroids.
I trained a DPLR NNP with a modifier. My Wannier centroid on O has a given charge, on the anion the number of Wannier centers averaged is not the same, the charge is different but this should not be a problem as the nuclei positive charges are also included in the modifier section of DeePMD input.

**DPLR MD with Lammps:**
I now have 3 real atom types: O, H and anion and 1 virtual atom (atom number 4 in Lammps data file) corresponding to the Wannier centroids on O and anion.
- I tried to have only one bond type for the WC-atom bond and 2 fix dplr in my Lammps input (see code snippet 1)
- I tried to have two different bond types for the WC-O and WC-anion bonds and 2 fix dplr in my Lammps input (see code snippet 2)
- I thought also that I could have only one bond type and 1 fix dplr but I did not manage to write a correct type_associate which should be something like `type_associate (1,3) 4 bond_type 1` but I cannot guess how to write it correctly

=> Nothing works, the two first tries output the same error: `ERROR: find a bonded pair the types of which are not associated (/home/conda/feedstock_root/build_artifacts/libdeepmd_1660884873849/work/source/lmp/fix_dplr.cpp:208)`

Code snippet 1:
```
fix             wcO all dplr model dplr.pb type_associate 1 4 bond_type 1 # for O - WC
fix_modify      wcO virial yes
fix             wcanion all dplr model dplr.pb type_associate 3 4 bond_type 1 # for anion - WC
fix_modify      wcanion virial yes
```

Code snippet 2:
```
fix             wcO all dplr model dplr.pb type_associate 1 4 bond_type 1 # for O - WC
fix_modify      wcO virial yes
fix             wcanion all dplr model dplr.pb type_associate 3 4 bond_type 2 # for anion - WC
fix_modify      wcanion virial yes
```

Do you have any experience with different atoms associated with their own Wannier centroid? From the fact that `model_charge_map` is a list in the DPLR model training json file it seems like it should be somehow possible. Can I have two `fix dplr` in my Lammps input? Should I have two bond types? Only one?

Thank you for your help!

**Comment by ChunyiZh** [comment, marked as answer]
Hi,

I would suggest that you set the Wannier centroids (WCs) for different atom types also as different types. For example, if your system is NaCl solution. You can set O as atom type 1, H as atom type 2, Na as atom type 3, Cl as atom type 4, the WCs of O as type 5, the WCs of Na as type 6, and the WCs of Cl as type 7. Then, there are 3 bond types, bond type 1 between atom types 1 and 5, bond type 2 between atom types 3 and 6, and bond type 3 between atom types 4 and 7. In Lammps input, just one "fix dplr" is enough.
 
Following are some settings for this system.

**Training DW model:**
                "fitting_net": {
                        "type": "dipole",
                        "sel_type": [
                                0,
                                2,
                                3
                        ],

**Training DPLR model:**
        "modifier": {
            "type":             "dipole_charge",
            "model_name":       "dw.pb",
            "model_charge_map": [-8,-8,-8],
            "sys_charge_map":   [6, 1, 9, 7],
            "ewald_h":          1.0,
            "ewald_beta":       0.40
        },

**Lammps input:**
neigh_modify    every 10 delay 0 check no exclude type 1 5 exclude type 3 6 exclude type 4 7
group           real_atom type 1 2 3 4
group           virtual_atom type 5 6 7
fix             0 all dplr model model_lr.pb type_associate 1 5 3 6 4 7 bond_type 1 2 3

**Reply by axgmz** [reply]
Thank you very much for this clear and detailed answer! It helped me a lot and solved my problem!
